etcdserver/server.go

```diff
@@ -1992,6 +1992,67 @@ func (s *EtcdServer) sync(timeout time.Duration) {
 	})
 }
 
+// publishV3 registers server information into the cluster using v3 request. The
+// information is the JSON representation of this server's member struct, updated
+// with the static clientURLs of the server.
+// The function keeps attempting to register until it succeeds,
+// or its server is stopped.
+func (s *EtcdServer) publishV3(timeout time.Duration) {
+	req := &membershippb.ClusterMemberAttrSetRequest{
+		Member_ID: uint64(s.id),
+		MemberAttributes: &membershippb.Attributes{
+			Name:       s.attributes.Name,
+			ClientUrls: s.attributes.ClientURLs,
+		},
+	}
+
+	for {
+		select {
+		case <-s.stopping:
+			if lg := s.getLogger(); lg != nil {
+				lg.Warn(
+					"stopped publish because server is stopping",
+					zap.String("local-member-id", s.ID().String()),
+					zap.String("local-member-attributes", fmt.Sprintf("%+v", s.attributes)),
+					zap.Duration("publish-timeout", timeout),
+				)
+			}
+			return
+
+		default:
+		}
+
+		ctx, cancel := context.WithTimeout(s.ctx, timeout)
+		_, err := s.raftRequest(ctx, pb.InternalRaftRequest{ClusterMemberAttrSet: req})
+		cancel()
+		switch err {
```

Comment:
Why removed `case ErrStopped`?

Reply:
My understanding is that `ErrStopped` is returned by `raftRequest()` when server closes `s.done` [1]. But the server waits for this loop to return before it closes `s.done` [2]. So they basically block each other. To solve this, I used `s.stopping` instead.

[1] https://github.com/YoyinZyc/etcd/blob/6fb95957bed41aca215f5a000ac4d25430be3984/etcdserver/v3_server.go#L663
[2] https://github.com/etcd-io/etcd/blob/18e9e45bda5818bae84be08585b13a7faf2263d3/etcdserver/server.go#L1012